extensions/amp-access-poool/0.1/poool-impl.js

```diff
@@ -0,0 +1,228 @@
+    return this.getPooolAccess_()
+        .then(response => {
+          return {access: response.access};
+        }, err => {
+          if (!err || !err.response) {
+            throw err;
+          }
+          const {response} = err;
+          if (response.status !== 402) {
+            throw err;
+          }
+          this.renderPoool_();
+          return {access: false};
+        });
+  }
+
+  /**
+   * @private
+   */
+  checkMandatoryParams_() {
+    user().assert(this.bundleID_, 'BundleID is incorrect or not provided.');
+    user().assert(this.pageType_, 'Page type is incorrect or not provided.');
+    user().assert(this.itemID_, 'ItemID is not provided.');
+  }
+
+  /**
+   * @return {!Promise<Object>}
+   * @private
+   */
+  getPooolAccess_() {
+    const url = this.accessUrl_ + '&iid=' + this.itemID_;
+    const urlPromise = this.accessSource_.buildUrl(url, false);
+    return urlPromise.then(url => {
+      return this.accessSource_.getLoginUrl(url);
+    }).then(url => {
+      dev().info(TAG, 'Authorization URL: ', url);
+      return this.timer_.timeoutPromise(
+          AUTHORIZATION_TIMEOUT,
+          this.xhr_.fetchJson(url)).then(res => res.json());
+    });
+  }
+
+  /**
+   * @private
+   */
+  renderPoool_() {
+    this.getPooolSDK_(global, _poool => {
+
+      // Init poool
+      _poool('init', this.bundleID_);
+
+      // Get AMP reader id and assign it to poool config
+      this.accessSource_.getReaderId_().then(rid => {
+        this.readerID_ = rid;
+
+        // Set poool amp basic config
+        _poool(
+            'config',
+            {
+              mode: 'custom',
+              'amp_reader_id': this.readerID_,
+              'amp_item_id': this.itemID_,
+            },
+            true
+        );
+      });
+
+
+      // Set config
+      Object.entries(CONFIG).forEach(configEntry => {
+
+        const configKey = configEntry[0],
+            configDefaultValue = configEntry[1];
+
+        let configValue = this.pooolConfig_[configKey] || configDefaultValue;
+
+        if (configValue) {
+          if (typeof configDefaultValue === 'boolean') {
+            configValue = `${configValue}` === 'true';
+          }
+
+          _poool('config', dashToUnderline(camelCaseToDash(configKey)),
+              configValue);
+        }
+      });
+
+      // Unlock content after onRelease event
+      _poool('event', 'onrelease', this.giveAccessOnPooolRelease_.bind(this));
+
+      // Create hit
+      _poool('send', 'page-view', this.pageType_);
+
+    });
+  }
+
+  /**
+   * @private
+   */
+  giveAccessOnPooolRelease_() {
+    const articleBody = document.getElementById('poool-access');
+    articleBody.setAttribute('amp-access-poool', '');
+  }
+
+  /**
+   * Produces the Poool SDK object for the passed in callback.
+   *
+   * @param {!Window} global
+   * @param {function(!Object)} cb
+   */
+  getPooolSDK_(global, cb) {
+    loadScript(global, 'https://assets.poool.fr/poool.min.js', function() {
```

Comment:
This is a problem. AMP does not allow 3p JS to be loaded on the main page.

Reply:
yeah we noticed that on travis, we will try a different approach, although our paywall is already executed inside an iframe.

Reply:
@dackmin To quickly add: there are couple of libraries that mostly operate via a 3p iframe. Please see `amp-access-*` and `amp-subscriptions-*` extensions.